lib/puppet/network/http.rb

```diff
@@ -5,6 +5,10 @@ module Puppet::Network::HTTP
   MASTER_URL_PREFIX = "/puppet"
   CA_URL_PREFIX = "/puppet-ca"
 
+  NOT_FOUND_ERROR_MESSAGE = 'This master is running Puppet 4, and the URL format has changed. ' +
+      'This request was likely made with a Puppet 3 Agent, as it used an invalid ' +
```

Comment:
Starting with puppet 3.3.1, we have `X-Puppet-Version` which we could use for more accurate reporting of the agent version.